### ISSUE TITLE
Document Machine Learning Algorithms Used in the Project

### ISSUE THREAD

**Issue by steev-vv** [steev-vv/frontend#1]
## Overview of Machine Learning Algorithms

The Smart Appointment System uses several machine learning algorithms to optimize different aspects of the appointment management process:

### 1. Gradient Boosting Classifier (GradientBoostingClassifier)
- **Used for**: No-show prediction in `NoShowPredictor` class
- **Why chosen**: Gradient Boosting is well-suited for binary classification problems like predicting patient no-shows because:
  - It handles imbalanced datasets effectively (no-shows are typically less common than attended appointments)
  - Provides good accuracy with interpretable feature importance
  - Performs well with a mix of categorical and numerical features
  - Robust against overfitting with appropriate hyperparameters

```python
self.model = GradientBoostingClassifier(
    n_estimators=100,
    learning_rate=0.1,
    max_depth=3,
    random_state=42
)
```

### 2. Random Forest Regressor (RandomForestRegressor)
- **Used for**: Wait time prediction in `WaitTimePredictor` class
- **Why chosen**: Random Forest is appropriate for wait time prediction because:
  - It handles non-linear relationships in the data effectively
  - Robust against outliers (some appointments may have unusually long wait times)
  - Ensemble method reduces variance and provides more stable predictions
  - Works well when multiple factors influence the outcome (doctor load, department load, time of day, etc.)

```python
self.model = RandomForestRegressor(
    n_estimators=100,
    max_depth=10,
    random_state=42
)
```

### 3. Custom Optimization Algorithm
- **Used for**: Schedule optimization in `ScheduleOptimizer` class
- **Implementation**: Custom weighted scoring system that considers:
  - Patient preferences
  - Appointment priorities
  - Current schedule constraints
  - Doctor availability

This algorithm calculates optimal schedules based on weighted factors rather than using a standard ML algorithm because scheduling is a constrained optimization problem that benefits from domain-specific logic.

### Recommended Improvements

1. Add hyperparameter tuning for the ML models
2. Consider implementing more advanced algorithms:
   - For no-show prediction: XGBoost or LightGBM for improved accuracy
   - For wait time prediction: Deep learning models for complex patterns
3. Add A/B testing framework to compare algorithm performance
4. Implement regular model retraining with new data

### Data Sources and Features

The models use various features including:
- Patient demographics and history
- Appointment time and day
- Department and doctor workload
- Previous appointment patterns
- Weather conditions (for no-show prediction)
